raspberry/main.py

```diff
@@ -0,0 +1,405 @@
+
+import numpy as np  # for matrix based calculations
+import pandas as pd  # pandas for storing cached data into csv files
+
+from time import time_ns  # function to get system-up time in nano seconds
+from datetime import datetime  # to get date-time stamps for naming the csv files
+from math import pi, atan2, sqrt  # math functions for calculations
+
+from bzzz.controllers.altitude_LQR import LQR  # altitude hold controller
+from bzzz.estimators.altitude_Kalman_filter import KalmanFilter
+from bzzz.scheduler import Scheduler
+from bzzz.sensors.time_of_flight_sensor import TimeOfFlightSensor
+from bzzz.read_sbus import RC  # for radio data receving, encoding and sending to ESP
+
+
+# NOTE: The scheduler supports both multi-threading and time-based function calling
+# Although threading guarantees consistent function call rates, the actual process handeling is not done
+# within the python environment which could cause unwanted behaviour.
+
+if __name__ == '__main__':
+    # sampling frequency of KF and LQR
+    sampling_frequency = 50
+    # caching configuration
+    enable_caching = [True]
+    enable_printing_cache_to_screen = [False and enable_caching[0]]
+
+    # objects declaration
+    kf = KalmanFilter(sampling_frequency=sampling_frequency,
+                      initial_Tt=0,
+                      x_tilde_0=np.array([[0], [0], [10], [-9.81]]),
+                      P_0=np.diagflat([1, 1, 1, 0.01]),
+                      cache_values=True)
+    lqr = LQR(sampling_frequency=sampling_frequency,
+              initial_alpha_t=10,
+              initial_beta_t=-9.81)
+    # update_measurement_at_fixed_rate: if True then use time difference between current time and last measurement time to take a measurement
+    #            if false then take a measurement instantly
+    tof = TimeOfFlightSensor(update_measurement_at_fixed_rate=False,
+                             median_filter_length=1,
+                             cache_altitude=True,
+                             use_outlier_detection=True,
+                             abs_outlier_diff_thres=500)
+    rc = RC()
+    scheduler = Scheduler(use_threading=False)
+
+    # NOTE: single element lists are used to avoid python-env re-declaring
+    # new local variables with in the functions that follow below.
+    # data caching and logging
+    time_cache = []
+    quat_cache = []
+    yaw_cache = []
+    pitch_cache = []
+    roll_cache = []
+    motor_PWM_cache = []
+    throttle_ref_cache = []
+    accelrometer_cache = []
+    altitude_reference_cache_mts = []
+    radio_data_cache = []
+    KF_data_cache = []
+    time_before_thread_starts = [0]
+
+    quaternion_vector = [0., 0., 0.]
+    acc = [0., 0., 0.]
+    euler = [0., 0., 0.]
+    motor_PWM = [0., 0., 0., 0.]  # FL, FR, BL, BR
+    channel_data = [""]
+    KF_data = [0., 0., 0., 0.]
+
+    switch_a_status = [True]
+    is_data_log_kill = [False]
+    is_kill = [False]
+    allow_data_logging = [True]
```

Comment:
Add a comment here or we'll soon forget what `allow_data_logging` is for

Reply:
Done!